dissect/cstruct/expression.py

```diff
@@ -1,84 +1,284 @@
+            right = self.queue.pop(-1)
+            res = ~right
+        else:
+            right = self.queue.pop(-1)
+            left = self.queue.pop(-1)
+            res = self.operators[operator](left, right)
+        self.queue.append(res)
 
-            buf += self.expression[i]
+    def isNumber(self, currentToken: str) -> bool:
+        return currentToken.isnumeric() or (
+            len(currentToken) > 2
+            and currentToken[0] == "0"
+            and (currentToken[1] == "x" or currentToken[1] == "X" or currentToken[1] == "b" or currentToken[1] == "o")
+        )
 
-        return self.evaluate_part(buf, context)
+    def parseErr(self, condition: bool, error: str) -> None:
+        if condition:
+            raise ExpressionParserError(error)
 
-    def evaluate_part(self, buf: str, context: Dict[str, int]) -> int:
-        buf = buf.strip()
+    def evaluate(self, context: Dict[str, int] = None) -> int:
+        """Evaluates an expression using a Shunting-Yard implementation"""
 
-        # Very simple way to support an expression(part) that is a single,
-        # negative value. To use negative values in more complex expressions,
-        # they must be wrapped in brackets, e.g.: 2 * (-5).
-        #
-        # To have full support for the negation operator a proper expression
-        # parser must be build.
-        if buf.startswith("-") and buf[1:].isnumeric():
-            return int(buf)
+        self.stack = []
+        self.queue = []
+        context = context or {}
+        tempExpression = self.tokens
+        opKeys = set(self.operators.keys())
 
-        for operator in self.operators:
-            if operator[0] in buf:
-                a, b = buf.rsplit(operator[0], 1)
+        # unary minus Tokens; we change the semantic of '-' depending on the previous token
+        for i in range(len(self.tokens)):
+            if self.tokens[i] == "-":
+                if i == 0:
+                    self.tokens[i] = "u"
+                    continue
+                if self.tokens[i - 1] in opKeys or self.tokens[i - 1] == "u" or self.tokens[i - 1] == "(":
+                    self.tokens[i] = "u"
+                    continue
 
-                return operator[1](self.evaluate_part(a, context), self.evaluate_part(b, context))
+        i = 0
+        while i < len(tempExpression):
+            currentToken = tempExpression[i]
+            if self.isNumber(currentToken):
+                self.queue.append(int(currentToken, 0))
+            elif currentToken in context:
+                self.queue.append(context[currentToken])
+            elif currentToken in self.cstruct.consts:
+                self.queue.append(self.cstruct.consts[currentToken])
+            elif currentToken == "u":
+                self.stack.append(currentToken)
+            elif currentToken == "~":
+                self.stack.append(currentToken)
+            elif currentToken == "sizeof":
+                assert tempExpression[i + 1] == "("
```

Comment:
Can you change these assert's to exceptions?

Assert's get removed when running Python with the `-O` flag.